File: C#/SubmodularFunction/Submodular/UnionFind.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

public class UnionFind
{
    int[] iarent;

    public UnionFind(int len)
    {
        iarent = new int[len];
        for (int i = 0; i < len; i++)
            iarent[i] = -1;
    }

    public int Length
    {
        get { return iarent.Length; }
    }

    public bool Unite(int x, int y)
    {
        x = Find(x);
        y = Find(y);
        if (x == y)
            return false;

        if (iarent[y] < iarent[x])    //swai
        {
            int tmp = x;
            x = y;
            y = tmp;
        }
        iarent[x] += iarent[y];
        iarent[y] = x;
        return true;
    }

    public int Find(int x)
    {
        if (iarent[x] < 0)
            return x;
        return iarent[x] = Find(iarent[x]);
    }

    public bool Same(int x, int y)
    {
        return Find(x) == Find(y);
    }

    public int Size(int x)
    {
        return -iarent[Find(x)];
    }

    public bool IsParent(int x)
    {
        return iarent[x] < 0;
    }

    public void Clear()
    {
        for (int i = 0; i < iarent.Length; i++)
            iarent[i] = -1;
    }

}

File: C++/SubmodularFunctionOld/SubmodularFunction/UnionFind.cpp
#include "Library.h"

using namespace OnigiriLibrary;


UnionFind::UnionFind(int len)
{
	UnionFind::len =len;
	parent = new int[len];
	for (int i = 0; i < len; i++)
		parent[i] = -1;
}

UnionFind::~UnionFind(){
	delete[] parent;
}

int UnionFind::Find(int x)
{
	if (parent[x] < 0)
		return x;
	return parent[x] = Find(parent[x]);
}

bool UnionFind::Unite(int x, int y)
{
	x = Find(x);
	y = Find(y);
	if (x == y)
		return false;

	if (parent[y] < parent[x])    //swap
	{
		int tmp = x;
		x = y;
		y = tmp;
	}
	parent[x] += parent[y];
	parent[y] = x;
	return true;
}

bool UnionFind::Same(int x, int y)
{
	return Find(x) == Find(y);
}

void UnionFind::Clear()
{
	for (int i = 0; i < len; i++)
		parent[i] = -1;
}
